### ISSUE TITLE
Should sparse-ho iteratively improve on MSE for test-data?

### ISSUE THREAD

**Issue by ksehic** [QB3/sparse-ho#33]
Hi,

I was running some wLasso examples with the initial step as the final estimation that you can get from LassoCV. The datasets were make_regression and 20news. I used the suggestion from @mathurinm to implement SNR, which is a nice way to see what is going on when noise is added gradually.

```
X, y = make_regression(n_samples=500, n_features=1000, noise=0)
noise_sample = np.random.randn(X.shape[0])
y = y + noise_level * noise_sample / np.linalg.norm(noise_sample) * np.linalg.norm(y)
```

I was expecting that sparse-ho should improve monotonically on validation loss without strange jumps that we can see on the plots. In addition, for every test that I made, MSE on test data was always increasing besides SNR=0.

To me, it is unclear how to specify max_epoich because you can easily get 
!!! Inner solver did not converge at epoch 9999, gap: 6.86e-02 > 1.19e-03

And if you put e.g. max_epoich = 1e9, some examples would run to inf. I changed wLasso for Lasso from sparse-ho for 20newsgroup and that was the case. This can happen for LassoCV from celer as well, but rare.

I would typically receive the warning related to line search !!step size rejected!!.

20newsgroup

![20news](https://user-images.githubusercontent.com/42025612/98790162-9c41ec80-2403-11eb-91fb-04c974981dac.png)

make_regression

![p1](https://user-images.githubusercontent.com/42025612/98791402-4d955200-2405-11eb-900d-a01816029086.png)

![p2](https://user-images.githubusercontent.com/42025612/98791417-51c16f80-2405-11eb-8e63-94c1b77b13a0.png)

The code I was running
```
# data select
X_train, X_val, X_test, y_train, y_val, y_test, X_train_val, y_train_val, train_indices, val_indices = data_select(
    data_pick="20news")#, noise_level=0.1)

# max penalty value
alpha_max = np.max(np.abs(X_train.T.dot(y_train))) / X_train.shape[0]

n_alphas = 30
alphas = alpha_max * np.geomspace(1, 0.001, n_alphas)

train_split_cv = [train_indices]
val_split_cv = [val_indices]


# vanilla LassoCV
print("========== Celer's LassoCV started ===============")
model_cv = LassoCV(verbose=False, fit_intercept=False,
                   alphas=alphas, tol=1e-5, max_iter=1e9,
                   cv=zip(train_split_cv, val_split_cv), max_epochs=1e9).fit(X_train_val,
                                                             y_train_val)

# measure mse on test
mse_cv = mean_squared_error(y_test,  model_cv.predict(X_test))
print("========== Celer's LassoCV finished ===============")
print("Vanilla LassoCV: Mean-squared error on test data %f" % mse_cv)

# let's run sparse-ho with vanilla lassoCV initial config
alpha0 = np.log(model_cv.alpha_) * np.ones(X_train.shape[1])
# alpha0 = np.log(model_cv.alpha_)
###
# Sparse-ho
lasso_sho = Lasso(fit_intercept=False, max_iter=1e9, warm_start=True, max_epochs=1e9)
model_sho = wLasso(X_train, y_train, estimator=lasso_sho)
# model_sho = sholasso(X_train, y_train, estimator=lasso_sho)
criterion_sho = CV(X_val, y_val, model_sho, X_test=X_test, y_test=y_test)
algo_sho = ImplicitForward(criterion_sho)
monitor = Monitor()
grad_search(
    algo_sho, alpha0, monitor, n_outer=20, tol=1e-5)

alphas = np.exp(monitor.log_alphas[-1])
# lasso_sho.fit(X_train / alphas, y_train)
lasso_sho.fit(X_train_val / alphas, y_train_val)
coef_ = lasso_sho.coef_ / alphas

mse_sho = mean_squared_error(y_test, X_test @ coef_)

print("========== Sparse-ho finished ===============")
print("Sparse-ho: Mean-squared error on test data %f" % mse_sho)

# plot mse in time
mse_sho_ite = np.zeros((20,))
for i in range(20):

    alphas = np.exp(monitor.log_alphas[i])
    # lasso_sho.fit(X_train / alphas, y_train)
    lasso_sho.fit(X_train_val / alphas, y_train_val)
    coef_ = lasso_sho.coef_ / alphas
    mse_sho_ite[i] = mean_squared_error(y_test, X_test @ coef_)

    print("Iter %d" % (i + 1))

# plot
fig = plt.figure()

plt.subplot(1,2,1)
plt.plot(np.linspace(1,20,20), monitor.objs, '-g',
         linewidth=3, label='Sparse-HO')
plt.scatter(np.linspace(1,20,20), monitor.objs, marker='o', c='g')
plt.ylabel("Validation Loss", fontsize=32)
plt.xlabel("Iteration", fontsize=32)
plt.rc('xtick', labelsize=32)
plt.rc('ytick', labelsize=32)
# plt.ylim(ymin=0)
plt.grid(True)
plt.legend(fontsize=32)

plt.subplot(1,2,2)
plt.plot(np.linspace(1,20,20), mse_sho_ite, '-g',
         linewidth=3, label='Sparse-HO')
plt.scatter(np.linspace(1,20,20), mse_sho_ite, marker='o', c='g')
plt.ylabel("MSE on test data", fontsize=32)
plt.xlabel("Iteration", fontsize=32)
plt.rc('xtick', labelsize=32)
plt.rc('ytick', labelsize=32)
# plt.ylim(ymin=0)
plt.grid(True)
plt.legend(fontsize=32)

plt.show()
```

**Comment by mathurinm** [comment]
Hello @ksehic, thanks for the feedback. 
In celer.LassoCV you can pass n_alphas=30 and eps=1e-3 to select 30 alphas from alpha max to alpha_max / 1000 (it makes your snippet shorter as you d'ont have to compute alpha_max etc, hence easier to parse) 

max_epochs=1e9 means maximum number of epochs is one billion, this may clearly lead to solver not returning anything. A good default is 50_000, maybe 100_000. If you get convergence errors with this setting, consider opening an issue on the celer repo.

I don't understand why, as SNR increases (snr is norm(X @ w_true) / norm(noise added), hence should be 1 / noise_level in your setting), the test error gets worse

**Comment by ksehic** [comment]
Hi, When I use max_epochs=1e5 for the synthetic data that you use to compare regression coefficients with the exact one [celer issue](https://github.com/mathurinm/celer/commit/582c74eaeda7547985fbcedcee7da0ff965ddc34), you can easily get "!!! Inner solver did not converge at epoch 99999, gap: 6.40e-05 > 6.01e-08" and it just keeps repeating itself. This happens when I use random search for alpha_range \in (np.log(alpha_max) np.log(alpha_max/10**4)). Meaning certain alpha configuration produces this error. I just wanted to let you know. 

Thanks, @mathurinm Sure, I will try and let you know asap.

**Comment by mathurinm** [comment]
My guess would be that the problem is really hard (large support, some values in alphas really close to 0). Going below alpha_max / 1e3 can take a really long time

**Comment by josephsalmon** [comment]
Could it be also normalization issues?
The gap (6.01e-08) targeted seems really small right?

**Comment by ksehic** [comment]
Yes, make sense.

**Comment by ksehic** [comment]
Hi, as @mathurinm suggested, I was running SNR on the normalized y and repeated the second plot on the original y. I see you have changed wLasso to weightedLasso, so I have updated my scripts.

So, the results are 

With the original y

![not_norm](https://user-images.githubusercontent.com/42025612/99673717-6c36c100-2a75-11eb-99e8-7e82f26817a0.png)

With the normalized y

![norm](https://user-images.githubusercontent.com/42025612/99673743-75279280-2a75-11eb-8420-cf42cf7a194c.png)

I can see that when SNR is increased, MSE is increased in both cases. However, I have noticed certain improvements such as MSE on the original y decreases for SNR 0.5 and 2. Maybe your new implementations within WeightedLasso has something to do with this.

I used max_epoich = 1e5 as suggested. The tolerance is 1e-5

Best

**Comment by mathurinm** [comment]
this tends to suggest numerical errors as pointed out by Joseph, no ? My first thought is that scaling y should not change the profile of the curves (for an appropriate scaling of alpha, we should get a scaled version of beta)
No ?

**Comment by QB3** [comment]
> I was expecting that sparse-ho should improve monotonically on validation loss without strange jumps that we can see on the plots. In addition

Unfortunately, we only provide a way to compute the gradient, then we combine this gradient scheme with a line-search procedure (https://github.com/fabianp/hoag/blob/6957c611aae01361786a2ce2bf12465086e9ecac/hoag/hoag.py#L202) which does not guaranty decreasing of the validation loss. A fortiori,t here are also no guarantees on the decreasing of the loss on the test set (i.e. on unseen data).

**Comment by QB3** [comment]
> To me, it is unclear how to specify max_epoich because you can easily get
!!! Inner solver did not converge at epoch 9999, gap: 6.86e-02 > 1.19e-03

This also a problem I have, when the stepsize is rejected, the procedure seems to fail.
I tried some other procedures, like ADAM: https://github.com/QB3/sparse-ho/blob/ce06e42aa4347c503d1ec4805eabdeb113f94826/sparse_ho/ho_dirty.py#L277 or coordinate descent scheme https://github.com/QB3/sparse-ho/blob/ce06e42aa4347c503d1ec4805eabdeb113f94826/sparse_ho/ho_dirty.py#L417 but it was not very succefull

**Comment by QB3** [comment]
> this tends to suggest numerical errors as pointed out by Joseph, no ? My first thought is that scaling y should not change the profile of the curves (for an appropriate scaling of alpha, we should get a scaled version of beta)
No ?

This may also be due to hardcoded values in the line search procedure: https://github.com/fabianp/hoag/blob/6957c611aae01361786a2ce2bf12465086e9ecac/hoag/hoag.py#L200

**Comment by QB3** [comment]
If you confirm everything works fine with the gradient descent, I will close the issue @ksehic
